### PR TITLE
Core support for enums, take #2

### PR DESCRIPTION
**Pull request by daniel-ohayon**
Second attempt to merge #14927 

This was previously reverted because of a performance regression detected in `TypeSignature.parseTypeSignature`

This regression is addressed in a newly added commit, which includes a benchmark.

Performance stats on BenchmarkTypeSignatureParsing are as follows on my machine:

```
parseRowTypeSignature
==================
* with enum parsing, unoptimized parsing: 600ms/op
* with enum parsing, optimized parsing: 2ms/op
* without enum parsing: 2ms/op

parseRowTypeSignatureWithEnums
==========================
* with enum parsing, optimized parsing: 2ms/op
```

### CODE REVIEW

**Comment by daniel-ohayon** [comment]
@rongrong I also added another new commit to the stack: Move enum parametric types to presto-common
This is so that plugins don't need to depend on presto-main to provide enum types.

I could also blend the last 3 commits into the existing commit but I figured that since the stack had already been merged into master and reverted, it might be clearer to leave the fixes as separate commits?

File: presto-common/src/test/java/com/facebook/presto/common/type/BenchmarkTypeSignatureParsing.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
...13 row(field_20 varchar, field_137 varchar, field_1 bigint, field_92 bigint, field_166 row(field_246 row(field_155 varchar), field_7 row(field_234 varchar, field_85 varchar, field_244 varchar, field_72 varchar, field_248 bigint, field_21 varchar, field_172 bigint, field_205 bigint, field_222 bigint), field_146 row(field_74 bigint, field_142 bigint, field_242 bigint, field_155 varchar, field_97 row(field_9 varchar), field_89 bigint), field_61 row(field_74 bigint, field_127 boolean, field_155 varchar)), field_175 array(varchar)), field_26 array(row(field_92 bigint, field_60 varchar)), field_240 row(field_233 boolean, field_119 boolean, field_57 bigint, field_162 bigint, field_14 boolean, field_4 double, field_37 double), field_224 row(field_233 boolean, field_119 boolean, field_57 bigint, field_162 bigint, field_34 boolean, field_207 boolean, field_65 boolean, field_28 bigint), field_43 row(field_215 varchar, field_43 varchar), field_101 row(field_233 boolean, field_119 boolean, field_57 bigint, field_162 bigint, field_14 boolean, field_4 double, field_37 double), field_29 row(field_151 row(field_92 bigint, field_235 varchar), field_135 varchar, field_92 bigint), field_75 boolean))";
```
Comment (rongrong) [inline presto-common/src/test/java/com/facebook/presto/common/type/BenchmarkTypeSignatureParsing.java:57]: Can we also add test case for signature with enums?

**Comment by rongrong** [comment]
> @rongrong I also added another new commit to the stack: Move enum parametric types to presto-common
> This is so that plugins don't need to depend on presto-main to provide enum types.
> 
> I could also blend these last 3 "fix" commits into the earlier commits but I figured that since the stack had already been merged into master and reverted, it might be clearer to leave the fixes as separate commits?

We already reverted the previous commits so let's just try to go clean and make each commit as bug-free as possible. It's easier to review the changes this way, but when we are ready to merge, please merge the new commits with the original ones introducing the change.

File: presto-common/src/test/java/com/facebook/presto/common/type/BenchmarkTypeSignatureParsing.java (hunk 1 above)
Comment (daniel-ohayon) [reply presto-common/src/test/java/com/facebook/presto/common/type/BenchmarkTypeSignatureParsing.java:64]: added

**Comment by daniel-ohayon** [comment]
OK, I combined the last 3 commits into the earlier ones (that includes your fix to the TreeRewriter)

**Comment by rongrong** [comment]
@yingsu00 @caithagoras We fixed the regression and added a benchmark to test `parseTypeSignature`. What do you guys fill about this? Should we do a verifier test before merging or rely on the nightly and release verification? I guess I'm the next release oncall so might be alright to just merge now? 😂

File: presto-common/src/test/java/com/facebook/presto/common/type/BenchmarkTypeSignatureParsing.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+@Fork(1)
+@Warmup(iterations = 5, timeUnit = TimeUnit.MILLISECONDS)
+@Measurement(iterations = 10, timeUnit = TimeUnit.MILLISECONDS)
+@BenchmarkMode(Mode.AverageTime)
+public class BenchmarkTypeSignatureParsing
+{
+    @Benchmark
+    public void parseRowTypeSignature(BenchmarkData data)
+    {
+        TypeSignature.parseTypeSignature(data.signature);
```
Comment (rongrong) [inline presto-common/src/test/java/com/facebook/presto/common/type/BenchmarkTypeSignatureParsing.java:45]: Static import. In general our style convention is to always do static import unless it's too ambiguous.

**Review by rongrong — APPROVED** [review]

File: presto-common/src/test/java/com/facebook/presto/common/type/BenchmarkTypeSignatureParsing.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    @Benchmark
```
Comment (rongrong) [inline presto-common/src/test/java/com/facebook/presto/common/type/BenchmarkTypeSignatureParsing.java:48]: Can you update the description with benchmark results of both tests? Thanks!

**Pull request merged**